restorers/evaluation/base.py

```diff
@@ -0,0 +1,190 @@
+from abc import ABC, abstractmethod
+from collections import OrderedDict
+from typing import List, Dict, Callable, Optional
+
+import wandb
+import numpy as np
+import tensorflow as tf
+from tqdm.auto import tqdm
+
+from ..utils import fetch_wandb_artifact
+
+
+class BaseEvaluator(ABC):
+    def __init__(
+        self, metrics: Dict[str, Callable], model: Optional[tf.keras.Model] = None
+    ):
+        """Base Class for Evaluating an Image Restoration Model"""
+        self.model = model
+        self.metrics = OrderedDict(metrics)
+        self.evaluation_report = {}
+        self.image_paths = self.populate_image_paths()
+        self.wandb_table = self.create_wandb_table() if wandb.run is not None else None
+
+    @abstractmethod
+    def preprocess(self, image_path: str):
+        raise NotImplementedError(f"{self.__class__.__name__ }.preprocess")
+
+    @abstractmethod
+    def postprocess(self, input_tensor: np.array):
+        raise NotImplementedError(f"{self.__class__.__name__ }.postprocess")
+
+    @abstractmethod
+    def populate_image_paths(self) -> Dict[str, Dict[str, List[str]]]:
+        raise NotImplementedError(f"{self.__class__.__name__ }.populate_image_paths")
+
+    def create_wandb_table(self) -> wandb.Table:
+        columns = ["Split", "Input-Image", "Ground-Truth-Image", "Enhanced-Image"]
+        columns = columns + [metric_alias for metric_alias, _ in self.metrics.items()]
+        return wandb.Table(columns=columns)
+
+    def initialize_model_from_wandb_artifact(self, artifact_address: str) -> None:
+        model_path = fetch_wandb_artifact(artifact_address, artifact_type="model")
+        self.model = tf.keras.models.load_model(model_path, compile=False)
+
+    def count_params(self, weights) -> int:
+        """Count the total number of scalars composing the weights.
+
+        Reference:
+            https://github.com/keras-team/keras/blob/e6784e4302c7b8cd116b74a784f4b78d60e83c26/keras/utils/layer_utils.py#L107
+
+        Args:
+            weights: An iterable containing the weights on which to compute params
+
+        Returns:
+            (int): The total number of scalars composing the weights
+        """
+        unique_weights = {id(w): w for w in weights}.values()
+        # Ignore TrackableWeightHandlers, which will not have a shape defined.
+        unique_weights = [w for w in unique_weights if hasattr(w, "shape")]
+        weight_shapes = [w.shape.as_list() for w in unique_weights]
+        standardized_weight_shapes = [
+            [0 if w_i is None else w_i for w_i in w] for w in weight_shapes
+        ]
+        return int(sum(np.prod(p) for p in standardized_weight_shapes))
+
+    def get_gflops(self) -> float:
```

Comment:
@ariG23498 @SauravMaheshkar The evaluation infrastructure is ready. However, I was unable to get this particular function working. Can you guys please lend some help?

Workspace for reference: https://wandb.ai/ml-colabs/zero-dce